File: 1-serverless-ts/cdk/config-tracker.stack.ts
import * as cdk from 'aws-cdk-lib';
import { CfnElement, CfnOutput, Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { ApiNestedStack } from './api.nested-stack';
import { NotificationsNestedStack } from './notifications.nested-stack';
import { StorageNestedStack } from './storage.nested-stack';

export class ConfigTrackerStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const { configTable, configChangesTopic } = new StorageNestedStack(this, 'Storage', {});
    const { api, apiKey } = new ApiNestedStack(this, 'Api', { configTable });
    new NotificationsNestedStack(this, 'Notifications', { configChangesTopic });

    // Outputs
    new CfnOutput(this, 'RestApiUrl', { value: api.url });
    new CfnOutput(this, 'RestApiKeyCommand', { value: `aws apigateway get-api-key --api-key ${apiKey.keyId} --include-value --query 'value' --output text` });
  }

  // https://github.com/aws/aws-cdk/issues/19099
  getLogicalId(element: CfnElement): string {
    if (element.node.id.includes('NestedStackResource')) {
      return /([a-zA-Z0-9]+)\.NestedStackResource/.exec(element.node.id)![1] // will be the exact id of the stack
    }
    return super.getLogicalId(element)
  }
}

File: 2-serverless-java/src/main/java/com/github/moravcik/configtracker/cdk/ConfigTrackerStack.java
package com.github.moravcik.configtracker.cdk;

import org.jetbrains.annotations.NotNull;
import software.amazon.awscdk.CfnElement;
import software.amazon.awscdk.StackProps;
import software.constructs.Construct;
import software.amazon.awscdk.Stack;
import software.amazon.awscdk.NestedStackProps;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class ConfigTrackerStack extends Stack {
    public ConfigTrackerStack(final Construct scope, final String id) {
        this(scope, id, null);
    }

    public ConfigTrackerStack(final Construct scope, final String id, final StackProps props) {
        super(scope, id, props);

        StorageNestedStack storageStack = new StorageNestedStack(this, "Storage");
        ApiNestedStack apiStack = new ApiNestedStack(this, "Api", storageStack.getConfigTable());
        new NotificationsNestedStack(this, "Notifications", storageStack.getConfigChangesTopic());

        // Outputs
        software.amazon.awscdk.CfnOutput.Builder.create(this, "RestApiUrl")
                .value(apiStack.getApiUrl())
                .build();
        software.amazon.awscdk.CfnOutput.Builder.create(this, "RestApiKeyCommand")
                .value("aws apigateway get-api-key --api-key " + apiStack.getApiKeyId() + " --include-value  --query 'value' --output text")
                .build();
    }

    @Override
    public @NotNull String getLogicalId(@NotNull CfnElement element) {
        if (element.getNode().getId().contains("NestedStackResource")) {
            Pattern pattern = Pattern.compile("([a-zA-Z0-9]+)\\.NestedStackResource");
            Matcher matcher = pattern.matcher(element.getNode().getId());
            if (matcher.find()) {
                return matcher.group(1);
            }
        }
        return super.getLogicalId(element);
    }
}
